libkbfs/bare_root_metadata_v2.go

```diff
@@ -1136,143 +1061,224 @@ func (md *BareRootMetadataV2) GetCurrentTLFPublicKey(
+	crypto cryptoPure, keyGen KeyGen,
+	updatedReaderKeys UserDevicePublicKeys,
+	ePubKey kbfscrypto.TLFEphemeralPublicKey,
+	ePrivKey kbfscrypto.TLFEphemeralPrivateKey,
+	tlfCryptKey kbfscrypto.TLFCryptKey) (UserDeviceKeyServerHalves, error) {
+	if len(updatedReaderKeys) != 1 {
+		return nil, fmt.Errorf("Expected updatedReaderKeys to have only one entry, got %d",
+			len(updatedReaderKeys))
 	}
-	ok, err := kbfscodec.Equal(codec, md.WKeys, md2.WKeys)
+
+	_, rkb, err := md.getTLFKeyBundles(keyGen)
 	if err != nil {
-		return false, err
-	}
-	if !ok {
-		return false, nil
+		return nil, err
 	}
-	ok, err = kbfscodec.Equal(codec, md.RKeys, md2.RKeys)
+
+	// This is VERY ugly, but we need it in order to avoid
+	// having to version the metadata. The index will be
+	// strictly negative for reader ephemeral public keys.
+	newIndex := -len(rkb.TLFReaderEphemeralPublicKeys) - 1
+
+	rServerHalves, err := rkb.RKeys.fillInUserInfos(
+		crypto, newIndex, updatedReaderKeys, ePrivKey, tlfCryptKey)
 	if err != nil {
-		return false, err
+		return nil, err
 	}
-	return ok, nil
-}
 
-// GetUnresolvedParticipants implements the BareRootMetadata interface for BareRootMetadataV2.
-func (md *BareRootMetadataV2) GetUnresolvedParticipants() (readers, writers []keybase1.SocialAssertion) {
-	return md.UnresolvedReaders, md.WriterMetadataV2.Extra.UnresolvedWriters
+	if len(rServerHalves) > 0 {
+		rkb.TLFReaderEphemeralPublicKeys = append(
+			rkb.TLFReaderEphemeralPublicKeys, ePubKey)
+	}
+
+	return rServerHalves, nil
 }
 
-// GetUserDeviceKeyInfoMaps implements the BareRootMetadata interface for BareRootMetadataV2.
-func (md *BareRootMetadataV2) GetUserDeviceKeyInfoMaps(
-	codec kbfscodec.Codec, keyGen KeyGen, _ ExtraMetadata) (
-	readers, writers UserDeviceKeyInfoMap, err error) {
+func (md *BareRootMetadataV2) updateKeyGeneration(
+	crypto cryptoPure, keyGen KeyGen,
+	updatedWriterKeys, updatedReaderKeys UserDevicePublicKeys,
+	ePubKey kbfscrypto.TLFEphemeralPublicKey,
+	ePrivKey kbfscrypto.TLFEphemeralPrivateKey,
+	tlfCryptKey kbfscrypto.TLFCryptKey) (UserDeviceKeyServerHalves, error) {
+	if len(updatedWriterKeys) == 0 {
+		return nil, errors.New(
+			"updatedWriterKeys unexpectedly non-empty")
+	}
+
 	wkb, rkb, err := md.getTLFKeyBundles(keyGen)
 	if err != nil {
-		return nil, nil, err
+		return nil, err
 	}
 
-	rUDKIM, err := rkb.RKeys.toUDKIM(codec)
+	newIndex := len(wkb.TLFEphemeralPublicKeys)
+
+	wServerHalves, err := wkb.WKeys.fillInUserInfos(
+		crypto, newIndex, updatedWriterKeys, ePrivKey, tlfCryptKey)
 	if err != nil {
-		return nil, nil, err
+		return nil, err
 	}
 
-	wUDKIM, err := wkb.WKeys.toUDKIM(codec)
+	rServerHalves, err := rkb.RKeys.fillInUserInfos(
+		crypto, newIndex, updatedReaderKeys, ePrivKey, tlfCryptKey)
 	if err != nil {
-		return nil, nil, err
+		return nil, err
 	}
 
-	return rUDKIM, wUDKIM, nil
+	serverHalves, err := wServerHalves.mergeUsers(rServerHalves)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(serverHalves) > 0 {
```

Comment:
Added comment. If we didn't update with any new devices, we shouldn't add an (unused) ephemeral key.